From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nicolas Iooss <nicolas.iooss_linux@m4x.org>
Date: Tue, 16 Aug 2016 11:07:13 +0200
Subject: [PATCH] {For LLVMLinux} Add an ugly workaround for clang+binutils
 2.27+debug build setup

Work around assembler error (clang is using gas as its integrated
assembler does not provide some GNU-specific extensions used in the
kernel):

    Error: inconsistent uses of .cfi_sections

It has been reported upstream in:
https://groups.google.com/forum/#!topic/llvm-dev/vgVAb2B9DVg
---
 Makefile                       | 2 ++
 clang-cfi-sections-directive.h | 7 +++++++
 2 files changed, 9 insertions(+)
 create mode 100644 clang-cfi-sections-directive.h

diff --git a/Makefile b/Makefile
index ead5e0b37c46..de1057450a3f 100644
--- a/Makefile
+++ b/Makefile
@@ -712,6 +712,8 @@ endif
 KBUILD_CFLAGS += $(CLANG_TARGET) $(CLANG_GCC_TC)
 KBUILD_AFLAGS += $(CLANG_TARGET) $(CLANG_GCC_TC)
 KBUILD_CPPFLAGS += $(call cc-option,-Qunused-arguments,)
+#KBUILD_CPPFLAGS += $(call cc-option,-Wno-unknown-warning-option,)
+KBUILD_CFLAGS += -include $(srctree)/clang-cfi-sections-directive.h
 #KBUILD_CFLAGS += $(call cc-disable-warning, unused-variable)
 #KBUILD_CFLAGS += $(call cc-disable-warning, format-invalid-specifier)
 #KBUILD_CFLAGS += $(call cc-disable-warning, gnu)
diff --git a/clang-cfi-sections-directive.h b/clang-cfi-sections-directive.h
new file mode 100644
index 000000000000..a9b301114918
--- /dev/null
+++ b/clang-cfi-sections-directive.h
@@ -0,0 +1,7 @@
+/* binutils 2.27 does not like the .cfi_sections directive generated by clang with debugging option.
+ * As a workaround, force .cfi_sections generation at the beginning of the file
+ */
+#ifndef __clang__
+#error This file is for clang only
+#endif
+asm(".cfi_sections .debug_frame\n");
-- 
